File: 3. Behavioral Design Patterns/State/C#/StateExample.cs
// State Pattern Example in C#
// TrafficLight cycles through Red, Yellow, Green states

using System;

namespace StateExample
{
    interface ITrafficLightState { void Next(TrafficLight context); string GetColor(); }

    class TrafficLight
    {
        private ITrafficLightState _state;
        public TrafficLight() { _state = new RedState(); }
        public void SetState(ITrafficLightState state) { _state = state; }
        public void Next() { _state.Next(this); }
        public string GetColor() => _state.GetColor();
    }

    class RedState : ITrafficLightState
    {
        public void Next(TrafficLight context) { context.SetState(new GreenState()); }
        public string GetColor() => "Red";
    }

    class YellowState : ITrafficLightState
    {
        public void Next(TrafficLight context) { context.SetState(new RedState()); }
        public string GetColor() => "Yellow";
    }

    class GreenState : ITrafficLightState
    {
        public void Next(TrafficLight context) { context.SetState(new YellowState()); }
        public string GetColor() => "Green";
    }

    class Program
    {
        static void Main()
        {
            var light = new TrafficLight();
            for (int i = 0; i < 6; i++)
            {
                Console.WriteLine($"Light: {light.GetColor()}");
                light.Next();
            }
        }
    }
}

File: 3. Behavioral Design Patterns/State/Java/StateExample.java
// State Pattern Example in Java
// TrafficLight cycles through Red, Yellow, Green states

interface TrafficLightState { void next(TrafficLight context); String getColor(); }

class TrafficLight {
    private TrafficLightState state;
    TrafficLight(){ state = new RedState(); }
    void setState(TrafficLightState s){ state = s; }
    void next(){ state.next(this); }
    String getColor(){ return state.getColor(); }
}

class RedState implements TrafficLightState { public void next(TrafficLight t){ t.setState(new GreenState()); } public String getColor(){ return "Red"; } }
class YellowState implements TrafficLightState { public void next(TrafficLight t){ t.setState(new RedState()); } public String getColor(){ return "Yellow"; } }
class GreenState implements TrafficLightState { public void next(TrafficLight t){ t.setState(new YellowState()); } public String getColor(){ return "Green"; } }

public class StateExample {
    public static void main(String[] args){
        TrafficLight light = new TrafficLight();
        for(int i=0;i<6;i++){ System.out.println("Light: " + light.getColor()); light.next(); }
    }
}
